### ISSUE TITLE
Machine Learning for Everyone :: In simple words. With real-world examples. Yes again :: vas3k's blog

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#2695]
Machine Learning for Everyone :: In simple words. With real-world examples. Yes, again :: vas3k's blog<br>
<br>
http://bit.ly/2De3ZT6<br>
<br>
Вастрик<br>
<br>
<div>
<div>
            <div>

<p>Machine Learning is like sex in high school. Everyone is talking about it, a few know what to do, and only your teacher is doing it. If you ever tried to read articles about machine learning on the Internet, most likely you stumbled upon two types of them: thick academic trilogies filled with theorems (I couldn’t even get through half of one) or fishy fairytales about <em>artificial intelligence</em>, <em>data-science magic</em>, and <em>jobs of the future</em>.</p>

<p>I decided to write a post I’ve been wishing existed for a long time. A simple introduction for those who always wanted to understand machine learning. Only real-world problems, practical solutions, simple language, and no high-level theorems. One and for everyone. Whether you are a programmer or a manager.</p>

<p>Let's roll.</p>

</div>

<div>
<h2>Why do we want machines to learn?</h2>

<p>This is Billy. Billy wants to buy a car. He tries to calculate how much he needs to save monthly for that. He went over dozens of ads on the internet and learned that new cars are around $20,000, used year-old ones are $19,000, 2-year old are $18,000 and so on.</p>

<p>Billy, our brilliant analytic, starts seeing a pattern: <em>so, the car price depends on its age and drops $1,000 every year, but won't get lower than $10,000.</em></p>

<p>In machine learning terms, Billy invented <em>regression</em> – he predicted a value (price) based on known historical data. People do it all the time, when trying to estimate a reasonable cost for a used iPhone on eBay or figure out how many ribs to buy for a BBQ party. 200 grams per person? 500?</p>

<p>Yeah, it would be nice to have a simple formula for every problem in the world. Especially, for a BBQ party. Unfortunately, it's impossible.</p>

<p>Let's get back to cars. The problem is, they have different manufacturing dates, dozens of options, technical condition, seasonal demand spikes, and god only knows how many more hidden factors. An average Billy can't keep all that data in his head while calculating the price. Me too.</p>

<p>People are dumb and lazy – we need robots to do the maths for them. So, let's go the computational way here. Let's provide the machine some data and ask it to find all hidden patterns related to price.</p>

<p>Aaaand it works. The most exciting thing is that the machine copes with this task much better than a real person does when carefully analyzing all the dependencies in their mind.</p>

<p>That was the birth of machine learning.</p>

</div>

<div>
<h2>Three components of machine learning</h2>

<p>Without all the AI-bullshit, the only goal of machine learning is to predict results based on incoming data. That's it. All ML tasks can be represented this way, or it's not an ML problem from the beginning.</p>

<p>The greater variety in the samples you have, the easier it is to find relevant patterns and predict the result. Therefore, we need three components to teach the machine:</p>

<p><span>Data</span> Want to detect spam? Get samples of spam messages. Want to forecast stocks? Find the price history. Want to find out user preferences? Parse their activities on Facebook (no, Mark, stop collecting it, enough!). The more diverse the data, the better the result. Tens of thousands of rows is the bare minimum for the desperate ones.</p>

<p>There are two main ways to get the data — manual and automatic. Manually collected data contains far fewer errors but takes more time to collect — that makes it more expensive in general.</p>

<p>Automatic approach is cheaper — you're gathering everything you can find and hope for the best.</p>

<p>Some smart asses like Google use their own customers to label data for them for free. Remember ReCaptcha which forces you to "Select all street signs"? That's exactly what they're doing. Free labour! Nice. In their place, I'd start to show captcha more and more. Oh, wait...</p>

<p>It's extremely tough to collect a good collection of data (usually called a dataset). They are so important that companies may even reveal their algorithms, but rarely datasets.</p>

<p><span>Features</span> Also known as parameters or variables. Those could be car mileage, user's gender, stock price, word frequency in the text. In other words, these are the factors for a machine to look at.</p>

<p>When data stored in tables it's simple — features are column names. But what are they if you have 100 Gb of cat pics? We cannot consider each pixel as a feature. That's why selecting the right features usually takes way longer than all the other ML parts. That's also the main source of errors. Meatbags are always subjective. They choose only features they like or find "more important". Please, avoid being human.</p>

<p><span>Algorithms</span> Most obvious part. Any problem can be solved differently. The method you choose affects the precision, performance, and size of the final model. There is one important nuance though: if the data is crappy, even the best algorithm won't help. Sometimes it's referred as "garbage in – garbage out". So don't pay too much attention to the percentage of accuracy, try to acquire more data first.</p>

</div>

<div>
<h2>Learning vs Intelligence</h2>

<p>Once I saw an article titled <em>"Will neural networks replace machine learning?"</em> on some hipster media website. These media guys always call any shitty linear regression at least artificial intelligence, almost SkyNet. Here is a simple picture to show who is who.</p>

<p><strong>Artificial intelligence</strong> is the name of a whole knowledge field, similar to biology or chemistry.</p>

<p><strong>Machine Learning</strong> is a part of artificial intelligence. An important part, but not the only one.</p>

<p><strong>Neural Networks</strong> are one of machine learning types. A popular one, but there are other good guys in the class.</p>

<p><strong>Deep Learning</strong> is a modern method of building, training, and using neural networks. Basically, it's a new architecture. Nowadays in practice, no one separates deep learning from the "ordinary networks". We even use the same libraries for them. To not look like a dumbass, it's better just name the type of network and avoid buzzwords.</p>

<p>The general rule is to compare things on the same level. That's why the phrase <em>"will neural nets replace machine learning"</em> sounds like <em>"will the wheels replace cars"</em>. Dear media, it's compromising your reputation a lot.</p>
<table>
<thead><tr>
<th><strong>Machine can</strong></th>
<th><strong>Machine cannot</strong></th>
</tr></thead>
<tbody>
<tr>
<td><strong>Forecast</strong></td>
<td><strong>Create something new</strong></td>
</tr>
<tr>
<td><strong>Memorize</strong></td>
<td><strong>Get smart really fast</strong></td>
</tr>
<tr>
<td><strong>Reproduce</strong></td>
<td><strong>Go beyond their task</strong></td>
</tr>
<tr>
<td><strong>Choose best item</strong></td>
<td><strong>Kill all humans</strong></td>
</tr>
</tbody>
</table>
</div>

<h1>The map of the machine learning world</h1>
<div>
<p>If you are too lazy for long reads, take a look at the picture below to get some understanding.</p>

<p>Always important to remember — there is never a sole way to solve a problem in the machine learning world. There are always several algorithms that fit, and you have to choose which one fits better. Everything can be solved with a neural network, of course, but who will pay for all these GeForces?</p>

<p>Let's start with a basic overview. Nowadays there are four main directions in machine learning.</p>

</div>

<h1>Part 1. Classical Machine Learning</h1>
<div>
<p>The first methods came from pure statistics in the '50s. They solved formal math tasks — searching for patterns in numbers, evaluating the proximity of data points, and calculating vectors' directions.</p>

<p>Nowadays, half of the Internet is working on these algorithms. When you see a list of articles to "read next" or your bank blocks your card at random gas station in the middle of nowhere, most likely it's the work of one of those little guys.</p>

<p>Big tech companies are huge fans of neural networks. Obviously. For them, 2% accuracy is an additional 2 billion in revenue. But when you are small, it doesn't make sense. I heard stories of the teams spending a year on a new recommendation algorithm for their e-commerce website, before discovering that 99% of traffic came from search engines. Their algorithms were useless. Most users didn't even open the main page.</p>

<p>Despite the popularity, classical approaches are so natural that you could easily explain them to a toddler. They are like basic arithmetic — we use it every day, without even thinking.</p>

</div>

<div>
<h1>1.1 Supervised Learning</h1>

<p>Classical machine learning is often divided into two categories – <strong>Supervised</strong> and <strong>Unsupervised Learning</strong>.</p>

<p>In the first case, the machine has a "supervisor" or a "teacher" who gives the machine all the answers, like whether it's a cat in the picture or a dog. The teacher has already divided (labeled) the data into cats and dogs, and the machine is using these examples to learn. One by one. Dog by cat.</p>

<p>Unsupervised learning means the machine is left on its own with a pile of animal photos and a task to find out who's who. Data is not labeled, there's no teacher, the machine is trying to find any patterns on its own. We'll talk about these methods below.</p>

<p>Clearly, the machine will learn faster with a teacher, so it's more commonly used in real-life tasks. There are two types of such tasks: <strong>classification – an object's category prediction, and regression – prediction of a specific point on a numeric axis</strong>.</p>

</div>

<div>
<h2>Classification</h2>

<p><em>"Splits objects based at one of the attributes known beforehand. Separate socks by based on color, documents based on language, music by genre"</em></p>

<p>Today used for:<br>
– Spam filtering<br>
– Language detection<br>
– A search of similar documents<br>
– Sentiment analysis<br>
– Recognition of handwritten characters and numbers<br>
– Fraud detection</p>

<p>Popular algorithms: <a href="https://en.wikipedia.org/wiki/Naive_Bayes_classifier">Naive Bayes</a>, <a href="https://en.wikipedia.org/wiki/Decision_tree_learning">Decision Tree</a>, <a href="https://en.wikipedia.org/wiki/Logistic_regression">Logistic Regression</a>, <a href="https://en.wikipedia.org/wiki/K-nearest_neighbors_algorithm">K-Nearest Neighbours</a>, <a href="https://en.wikipedia.org/wiki/Support_vector_machine">Support Vector Machine</a></p>
<div>
<p>From here onward you can comment with additional information for these sections. Feel free to write your examples of tasks. Everything is written here based on my own subjective experience.</p>
</div>

</div>

<div>
<p>Machine learning is about classifying things, mostly. The machine here is like a baby learning to sort toys: here's a robot, here's a car, here's a robo-car... Oh, wait. Error! Error!</p>

<p>In classification, you always need a teacher. The data should be labeled with features so the machine could assign the classes based on them. Everything could be classified — users based on interests (as algorithmic feeds do), articles based on language and topic (that's important for search engines), music based on genre (Spotify playlists), and even your emails.</p>

<p>In spam filtering the <a href="https://www.analyticsvidhya.com/blog/2017/09/naive-bayes-explained/">Naive Bayes</a> algorithm was widely used. The machine counts the number of "viagra" mentions in spam and normal mail, then it multiplies both probabilities using the Bayes equation, sums the results and yay, we have Machine Learning.</p>

<p>Later, spammers learned how to deal with Bayesian filters by adding lots of "good" words at the end of the email. Ironically, the method was called <a href="https://en.wikipedia.org/wiki/Bayesian_poisoning">Bayesian poisoning</a>. Naive Bayes went down in history as the most elegant and first practically useful one, but now other algorithms are used for spam filtering.</p>

<p>Here's another practical example of classification. Let's say you need some money on credit. How will the bank know if you'll pay it back or not? There's no way to know for sure. But the bank has lots of profiles of people who took money before. They have data about age, education, occupation and salary and – most importantly – the fact of paying the money back. Or not.</p>

<p>Using this data, we can teach the machine to find the patterns and get the answer. There's no issue with getting an answer. The issue is that the bank can't blindly trust the machine answer. What if there's a system failure, hacker attack or <em>a quick fix from a drunk senior</em>.</p>

<p>To deal with it, we have <a href="https://www.youtube.com/watch?v=eKD5gxPPeY0">Decision Trees</a>. All the data automatically divided to yes/no questions. They could sound a bit weird from a human perspective, e.g., <em>whether the creditor earns more than $128.12?</em> Though, the machine comes up with such questions to split the data best at each step.</p>

<p>That's how a tree is made. The higher the branch — the broader the question. Any analyst can take it and explain afterward. He may not understand it, but explain easily! (typical analyst)</p>

<p>Decision trees are widely used in high responsibility spheres: diagnostics, medicine, and finances.</p>
<div>
<p>The two most popular algorithms for forming the trees are <a href="https://en.wikipedia.org/wiki/Predictive_analytics#Classification_and_regression_trees_.28CART.29">CART</a> and <a href="https://en.wikipedia.org/wiki/C4.5_algorithm">C4.5</a>.</p>
</div>

<p>Pure decision trees are rarely used today. However, they often set the basis for large systems, and their ensembles even work better than neural networks. We'll talk about that later.</p>
<div>
<p>When you google something, that's precisely the bunch of dumb trees which are looking for a range of answers for you. Search engines love them because they're fast.</p>
</div>

<p><em>Support Vector Machines (SVM)</em> is rightfully the most popular method of classical classification. It was used to classify everything in existence: plants by appearance in photos, documents by categories, etc.</p>

<p>The idea behind SVM is simple – it's trying to draw two lines between your data points with the largest margin between them. Look at the picture:</p>

<p>There's one very useful side of the classification — anomaly detection. When a feature does not fit any of the classes, we highlight it. Now that's used in medicine — on MRIs, computers highlight all the suspicious areas or deviations of the test. Stock markets use it to detect abnormal behaviour of traders to find the insiders. When teaching the computer the right things, we automatically teach it what things are wrong.</p>

<p>Today, neural networks are more frequently used for classification. Well, that's what they were created for.</p>

<p><strong>The rule of thumb is the more complex the data, the more complex the algorithm.</strong> For text, numbers, and tables, I'd choose the classical approach. The models are smaller there, they learn faster and work more clearly. For pictures, video and all other complicated big data things, I'd definitely look at neural networks.</p>

<p>Just five years ago you could find a face classifier built on SVM. Today it's easier to choose from hundreds of pre-trained networks. Nothing has changed for spam filters, though. They are still written with SVM. And there's no good reason to switch from it anywhere.</p>

<p>Even my website has SVM-based spam detection in comments ¯_(ツ)_/¯</p>

</div>

<div>
<h2>Regression</h2>

<p><em>"Draw a line through these dots. Yep, that's the machine learning"</em></p>

<p>Today this is used for:</p>
<ul>
<li>Stock price forecasts</li>
<li>Demand and sales volume analysis</li>
<li>Medical diagnosis</li>
<li>Any number-time correlations</li>
</ul>
<p>Popular algorithms are <a href="https://en.wikipedia.org/wiki/Linear_regression">Linear</a> and <a href="https://en.wikipedia.org/wiki/Polynomial_regression">Polynomial</a> regressions.</p>

</div>

<div>
<p>Regression is basically classification where we forecast a number instead of category. Examples are car price by its mileage, traffic by time of the day, demand volume by growth of the company etc. Regression is perfect when something depends on time.</p>

<p>Everyone who works with finance and analysis loves regression. It's even built-in to Excel. And it's super smooth inside — the machine simply tries to draw a line that indicates average correlation. Though, unlike a person with a pen and a whiteboard, machine does so with mathematical accuracy, calculating the average interval to every dot.</p>

<p>When the line is straight — it's a linear regression, when it's curved – polynomial. These are two major types of regression. The other ones are more exotic. Logistic regression is a black sheep in the flock. Don't let it trick you, as it's a classification method, not regression.</p>

<p>It's okay to mess with regression and classification, though. Many classifiers turn into regression after some tuning. We can not only define the class of the object but memorize how close it is. Here comes a regression.</p>

</div>

<div>
<h1>1.2 Unsupervised learning</h1>

<p>Unsupervised was invented a bit later, in the '90s. It is used less often, but sometimes we simply have no choice.</p>

<p>Labeled data is luxury. But what if I want to create, let's say, a bus classifier? Should I manually take photos of million fucking buses on the streets and label each of them? No way, that will take a lifetime, and I still have so many games not played on my Steam account.</p>

<p>There's a little hope for capitalism in this case. Thanks to social stratification, we have millions of cheap workers and services like <a href="https://www.mturk.com/">Mechanical Turk</a> who are ready to complete your task for $0.05. And that's how things usually get done here.</p>

<p>Or you can try to use unsupervised learning. But I can't remember any good practical application for it, though. It's usually useful for <a href="https://en.wikipedia.org/wiki/Exploratory_data_analysis">exploratory data analysis</a> but not as the main algorithm. Specially trained meatbag with Oxford degree feeds the machine with a ton of garbage and watches it. Are there any clusters? No. Any visible relations? No. Well, continue then. You wanted to work in data science, right?</p>

</div>

<div>
<h2>Clustering</h2>

<p><em>"Divides objects based on unknown features. Machine chooses the best way"</em></p>

<p>Nowadays used:</p>
<ul>
<li>For market segmentation (types of customers, loyalty)</li>
<li>To merge close points on a map</li>
<li>For image compression</li>
<li>To analyze and label new data</li>
<li>To detect abnormal behavior </li>
</ul>
<p>Popular algorithms: <a href="https://en.wikipedia.org/wiki/K-means_clustering">K-means_clustering</a>, <a href="https://en.wikipedia.org/wiki/Mean_shift">Mean-Shift</a>, <a href="https://en.wikipedia.org/wiki/DBSCAN">DBSCAN</a></p>

</div>

<div>
<p>Clustering is a classification with no predefined classes. It’s like dividing socks by color when you don't remember all the colors you have. Clustering algorithm trying to find similar (by some features) objects and merge them in a cluster. Those who have lots of similar features are joined in one class. With some algorithms, you even can specify the exact number of clusters you want.</p>

<p>An excellent example of clustering — markers on web maps. When you're looking for all vegan restaurants around, the clustering engine groups them to blobs with a number. Otherwise, your browser would freeze, trying to draw all three million vegan restaurants in that hipster downtown.</p>

<p>Apple Photos and Google Photos use more complex clustering. They're looking for faces in photos to create albums of your friends. The app doesn't know how many friends you have and how they look, but it's trying to find the common facial features. Typical clustering.</p>

<p>Another popular issue is image compression. When saving the image to PNG you can set the palette, let's say, to 32 colors. It means clustering will find all the "reddish" pixels, calculate the "average red" and set it for all the red pixels. Fewer colors — lower file size — profit!</p>

<p>However, you may have problems with colors like Cyan<span>◼︎</span>-like colors. Is it green or blue? Here comes the <a href="https://www.youtube.com/watch?v=_aWzGGNrcic">K-Means</a> algorithm.</p>

<p>It randomly sets 32 color dots in the palette. Now, those are centroids. The remaining points are marked as assigned to the nearest centroid. Thus, we get kind of galaxies around these 32 colors. Then we're moving the centroid to the center of its galaxy and repeat that until centroids stop moving.</p>

<p>All done. Clusters defined, stable, and there are exactly 32 of them. Here is a more real-world explanation:</p>

<p>Searching for the centroids is convenient. Though, in real life clusters not always circles. Let's imagine you're a geologist. And you need to find some similar minerals on the map. In that case, the clusters can be weirdly shaped and even nested. Also, you don't even know how many of them to expect. 10? 100?</p>

<p>K-means does not fit here, but <a href="https://towardsdatascience.com/how-dbscan-works-and-why-should-i-use-it-443b4a191c80">DBSCAN</a> can be helpful. Let's say, our dots are people at the town square. Find any three people standing close to each other and ask them to hold hands. Then, tell them to start grabbing hands of those neighbors they can reach. And so on, and so on until no one else can take anyone's hand. That's our first cluster. Repeat the process until everyone is clustered. Done.</p>
<div>
<p>A nice bonus: a person who has no one to hold hands with — is an anomaly.</p>
</div>

<p>It all looks cool in motion:</p>

<p>Just like classification, clustering could be used to detect anomalies. User behaves abnormally after signing up? Let the machine ban him temporarily and create a ticket for the support to check it. Maybe it's a bot. We don't even need to know what "normal behavior" is, we just upload all user actions to our model and let the machine decide if it's a "typical" user or not.</p>

<p>This approach doesn't work that well compared to the classification one, but it never hurts to try.</p>

</div>

<div>
<h2>Dimensionality Reduction (Generalization)</h2>

<p><em>"Assembles specific features into more high-level ones"</em></p>

<p>Nowadays is used for:</p>
<ul>
<li>Recommender systems (★)</li>
<li>Beautiful visualizations</li>
<li>Topic modeling and similar document search</li>
<li>Fake image analysis</li>
<li>Risk management</li>
</ul>
<p>Popular algorithms: <a href="https://en.wikipedia.org/wiki/Principal_component_analysis">Principal Component Analysis</a> (PCA), <a href="https://en.wikipedia.org/wiki/Singular_value_decomposition">Singular Value Decomposition</a> (SVD), <a href="https://en.wikipedia.org/wiki/Latent_Dirichlet_allocation">Latent Dirichlet allocation</a> (LDA), <a href="https://en.wikipedia.org/wiki/Probabilistic_latent_semantic_analysis">Latent Semantic Analysis</a> (LSA, pLSA, GLSA), <a href="https://en.wikipedia.org/wiki/T-distributed_stochastic_neighbor_embedding">t-SNE</a> (for visualization)</p>

</div>

<div>
<p>Previously these methods were used by hardcore data scientists, who had to find "something interesting" in huge piles of numbers. When Excel charts didn't help, they forced machines to do the pattern-finding. That's how they got Dimension Reduction or Feature Learning methods.</p>

<div>
<p><figure><a href="https://i.stack.imgur.com/Q7HIP.gif"><img src="https://i.stack.imgur.com/Q7HIP.gif"></a>
    <figcaption>Projecting 2D-data to a line (PCA)</figcaption></figure></p>
</div>

<p>It is always more convenient for people to use abstractions, not a bunch of fragmented features. For example, we can merge all dogs with triangle ears, long noses, and big tails to a nice abstraction — "shepherd". Yes, we're losing some information about the specific shepherds, but the new abstraction is much more useful for naming and explaining purposes. As a bonus, such "abstracted" models learn faster, overfit less and use a lower number of features.</p>

<p>These algorithms became an amazing tool for <span>Topic Modeling</span>. We can abstract from specific words to their meanings. This is what <a href="https://en.wikipedia.org/wiki/Latent_semantic_analysis">Latent semantic analysis</a> (LSA) does. It is based on how frequently you see the word on the exact topic. Like, there are more tech terms in tech articles, for sure. The names of politicians are mostly found in political news, etc.</p>

<p>Yes, we can just make clusters from all the words at the articles, but we will lose all the important connections (for example the same meaning of <em>battery</em> and <em>accumulator</em> in different documents). LSA will handle it properly, that's why its called "latent semantic".</p>

<p>So we need to connect the words and documents into one feature to keep these latent connections — it turns out that <a href="https://en.wikipedia.org/wiki/Singular_value_decomposition">Singular decomposition</a> (SVD) nails this task, revealing useful topic clusters from seen-together words.</p>

<p><span>Recommender Systems and Collaborative Filtering</span> is another super-popular use of the dimensionality reduction method. Seems like if you use it to abstract user ratings, you get a great system to recommend movies, music, games and whatever you want.</p>

<p>It's barely possible to fully understand this machine abstraction, but it's possible to see some correlations on a closer look. Some of them correlate with user's age — kids play Minecraft and watch cartoons more; others correlate with movie genre or user hobbies.</p>

<p>Machines get these high-level concepts even without understanding them, based only on knowledge of user ratings. Nicely done, Mr.Computer. Now we can write a thesis on why bearded lumberjacks love My Little Pony.</p>

</div>

<div>
<h2>Association rule learning</h2>

<p><em>"Look for patterns in the orders' stream"</em></p>

<p>Nowadays is used:</p>
<ul>
<li>To forecast sales and discounts</li>
<li>To analyze goods bought together</li>
<li>To place the products on the shelves</li>
<li>To analyze web surfing patterns</li>
</ul>
<p>Popular algorithms: <a href="https://en.wikipedia.org/wiki/Association_rule_learning#Algorithms">Apriori, Euclat, FP-growth</a></p>

</div>

<div>
<p>This includes all the methods to analyze shopping carts, automate marketing strategy, and other event-related tasks. When you have a sequence of something and want to find patterns in it — try these thingys.</p>

<p>Say, a customer takes a six-pack of beers and goes to the checkout. Should we place peanuts on the way? How often do people buy them together? Yes, it probably works for beer and peanuts, but what other sequences can we predict? Can a small change in the arrangement of goods lead to a significant increase in profits?</p>

<p>Same goes for e-commerce. The task is even more interesting there — what is the customer going to buy next time?</p>

<p>No idea why rule-learning seems to be the least elaborated upon category of machine learning. Classical methods are based on a head-on look through all the bought goods using trees or sets. Algorithms can only search for patterns, but cannot generalize or reproduce those on new examples.</p>

<p>In the real world, every big retailer builds their own proprietary solution, so nooo revolutions here for you. The highest level of tech here — recommender systems. Though, I may be not aware of a breakthrough in the area. Let me know in the comments if you have something to share.</p>

</div>

<h1>Part 2. Reinforcement Learning</h1>
<div>

<p><em>"Throw a robot into a maze and let it find an exit"</em></p>

<p>Nowadays used for:</p>
<ul>
<li>Self-driving cars</li>
<li>Robot vacuums</li>
<li>Games</li>
<li>Automating trading</li>
<li>Enterprise resource management</li>
</ul>
<p>Popular algorithms: <a href="https://en.wikipedia.org/wiki/Q-learning">Q-Learning</a>, <a href="https://en.wikipedia.org/wiki/State%E2%80%93action%E2%80%93reward%E2%80%93state%E2%80%93action">SARSA</a>, DQN, <a href="https://medium.com/emergent-future/simple-reinforcement-learning-with-tensorflow-part-8-asynchronous-actor-critic-agents-a3c-c88f72a5e9f2">A3C</a>, <a href="https://en.wikipedia.org/wiki/Genetic_algorithm">Genetic algorithm</a></p>

</div>

<div>
<p>Finally, we get to something looks like <em>real artificial intelligence</em>. In lots of articles reinforcement learning is placed somewhere in between of supervised and unsupervised learning. They have nothing in common! Is this because of the name?</p>

<p>Reinforcement learning is used in cases when your problem is not related to data at all, but you have an environment to live in. Like a video game world or a city for self-driving car.</p>
<p>
    VIDEO
    <small>Neural network plays Mario</small>
</p>
<p>Knowledge of all the road rules in the world will not teach the autopilot how to drive on the roads. Regardless of how much data we collect, we still can't foresee all the possible situations. This is why its goal is to <strong>minimize error, not to predict all the moves</strong>.</p>

<p>Surviving in an environment is a core idea of reinforcement learning. Throw poor little robot into real life, punish it for errors and reward it for right deeds. Same way we teach our kids, right?</p>

<p>More effective way here — to build a virtual city and let self-driving car to learn all its tricks there first. That's exactly how we train auto-pilots right now. Create a virtual city based on a real map, populate with pedestrians and let the car learn to kill as few people as possible. When the robot is reasonably confident in this artificial GTA, it's freed to test in the real streets. Fun!</p>

<p>There may be two different approaches — <strong>Model-Based and Model-Free</strong>.</p>

<p>Model-Based means that car needs to memorize a map or its parts. That's a pretty outdated approach since it's impossible for the poor self-driving car to memorize the whole planet.</p>

<p>In Model-Free learning, the car doesn't memorize every movement but tries to generalize situations and act rationally while obtaining a maximum reward.</p>

<p>Remember the news about <a href="https://www.wired.com/2016/01/in-a-huge-breakthrough-googles-ai-beats-a-top-player-at-the-game-of-go/">AI beating a top player at the game of Go</a>? Despite shortly before this it being <a href="https://motherboard.vice.com/en_us/article/vv7ejx/after-2500-years-a-chinese-gaming-mystery-is-solved">proved</a> that the number of combinations in this game is greater than the number of atoms in the universe.</p>

<p>This means the machine could not remember all the combinations and thereby win Go (as it did chess). At each turn, it simply chose the best move for each situation, and it did well enough to outplay a human meatbag.</p>

<p>This approach is a core concept behind <a href="https://www.youtube.com/watch?v=aCEvtRtNO-M">Q-learning</a> and its derivatives (SARSA &amp; DQN). 'Q' in the name stands for "Quality" as a robot learns to perform the most "qualitative" action in each situation and all the situations are memorized as a simple <a href="https://en.wikipedia.org/wiki/Markov_process">markovian process</a>.</p>

<p>Such a machine can test billions of situations in a virtual environment, remembering which solutions led to greater reward. But how can it distinguish previously seen situations from a completely new one? If a self-driving car is at a road crossing and the traffic light turns green — does it mean it can go now? What if there's an ambulance rushing through a street nearby?</p>

<p>The answer today is "no one knows". There's no easy answer. Researchers are constantly searching for it but meanwhile only finding workarounds. Some would hardcode all the situations manually that let them solve exceptional cases, like the <a href="https://en.wikipedia.org/wiki/Trolley_problem">trolley problem</a>. Others would go deep and let neural networks do the job of figuring it out. This led us to the evolution of Q-learning called Deep Q-Network (DQN). But they are not a silver bullet either.</p>

<p>Reinforcement Learning for an average person would look like a <em>real artificial intelligence</em>. Because it makes you think <em>wow, this machine is making decisions in real life situations</em>! This topic is hyped right now, it's advancing with incredible pace and intersecting with a neural network to clean your floor more accurately. Amazing world of technologies!</p>
<hr>
<p><strong>Off-topic.</strong> When I was a student, <a href="http://rednuht.org/genetic_walkers/">genetic algorithms</a> (link has cool visualization) were really popular. This is about throwing a bunch of robots into a single environment and making them try reaching the goal until they die. Then we pick the best ones, cross them, mutate some genes and rerun the simulation. After a few <a href="http://mathworld.wolfram.com/Milliard.html">milliard</a> years, we will get an intelligent creature. Probably. Evolution at its finest.</p>

<p>Genetic algorithms are considered as part of reinforcement learning and they have the most important feature proved by decade-long practice: no one gives a shit about them.</p>

<p>Humanity still couldn't come up with a task where those would be more effective than other methods. But they are great for student experiments and let people get their university supervisors excited about "artificial intelligence" without too much labour. And youtube would love it as well.</p>

</div>

<h1>Part 3. Ensemble Methods</h1>
<div>

<p><em>"Bunch of stupid trees learning to correct errors of each other"</em></p>

<p>Nowadays is used for:</p>
<ul>
<li>Everything that fits classical algorithm approaches (but works better)</li>
<li>Search systems (★)</li>
<li>Computer vision</li>
<li>Object detection</li>
</ul>
<p>Popular algorithms: <a href="https://en.wikipedia.org/wiki/Random_forest">Random Forest</a>, <a href="https://en.wikipedia.org/wiki/Gradient_boosting">Gradient Boosting</a></p>

</div>

<div>
<p>It's time for modern, grown-up methods. Ensembles and neural networks are two main fighters paving our path to a singularity. Today they are producing the most accurate results and are widely used in production.</p>

<p>However, the neural networks got all the hype today, while the words like "boosting" or "bagging" are scarce hipsters on TechCrunch.</p>

<p>Despite all the effectiveness the idea behind these is overly simple. If you take a bunch of inefficient algorithms and force them to correct each other's mistakes, the overall quality of a system will be higher than even the best individual algorithms.</p>

<p>You'll get even better results if you take the most unstable algorithms that are predicting completely different results on small noise in input data. Like Regression and Decision Trees. These algorithms are so sensitive to even a single outlier in input data to have models go mad.</p>

<p>In fact, this is what we need.</p>

<p>We can use any algorithm we know to create an ensemble. Just throw a bunch of classifiers, spice it up with regression and don't forget to measure accuracy. From my experience: don't even try a Bayes or kNN here. Although "dumb", they are really stable. That's boring and predictable. Like your ex.</p>

<p>Instead, there are three battle-tested methods to create ensembles.</p>

<p><span>Stacking</span> Output of several parallel models is passed as input to the last one which makes a final decision. Like that girl who asks her girlfriends whether to meet with you in order to make the final decision herself.</p>

<p>Emphasis here on the word "different". Mixing the same algorithms on the same data would make no sense. The choice of algorithms is completely up to you. However, for final decision-making model, regression is usually a good choice.</p>

<p>Based on my experience stacking is less popular in practice, because two other methods are giving better accuracy.</p>

<p><span>Bagging</span> aka <a href="http://en.wikipedia.org/wiki/Bootstrap_aggregating">Bootstrap AGGregatING</a>. Use the same algorithm but train it on different subsets of original data. In the end — just average answers.</p>

<p>Data in random subsets may repeat. For example, from a set like "1-2-3" we can get subsets like "2-2-3", "1-2-2", "3-1-2" and so on. We use these new datasets to teach the same algorithm several times and then predict the final answer via simple majority voting.</p>

<p>The most famous example of bagging is the <a href="https://www.youtube.com/watch?v=J4Wdy0Wc_xQ">Random Forest</a> algorithm, which is simply bagging on the decision trees (which were illustrated above). When you open your phone's camera app and see it drawing boxes around people's faces — it's probably the results of Random Forest work. Neural networks would be too slow to run real-time yet bagging is ideal given it can calculate trees on all the shaders of a video card or on these new fancy ML processors.</p>

<p>In some tasks, the ability of the Random Forest to run in parallel is more important than a small loss in accuracy to the boosting, for example. Especially in real-time processing. There is always a trade-off.</p>

<p><span>Boosting</span> Algorithms are trained one by one sequentially. Each subsequent one paying most of its attention to data points that were mispredicted by the previous one. Repeat until you are happy.</p>

<p>Same as in bagging, we use subsets of our data but this time they are not randomly generated. Now, in each subsample we take a part of the data the previous algorithm failed to process. Thus, we make a new algorithm learn to fix the errors of the previous one.</p>

<p>The main advantage here — a very high, even illegal in some countries precision of classification that all cool kids can envy. The cons were already called out — it doesn't parallelize. But it's still faster than neural networks. It's like a race between a dump truck and a racecar. The truck can do more, but if you want to go fast — take a car.</p>

<p>If you want a real example of boosting — open Facebook or Google and start typing in a search query. Can you hear an army of trees roaring and smashing together to sort results by relevancy? That's because <a href="https://code.fb.com/ml-applications/evaluating-boosted-decision-trees-for-billions-of-users/">they are using boosting</a>.</p>

</div>

<h1>Part 4. Neural Networks and Deep Leaning</h1>
<div>

<p><em>"We have a thousand-layer network, dozens of video cards, but still no idea where to use it. Let's generate cat pics!"</em></p>

<p>Used today for:</p>
<ul>
<li>Replacement of all algorithms above</li>
<li>Object identification on photos and videos</li>
<li>Speech recognition and synthesis</li>
<li>Image processing, style transfer</li>
<li>Machine translation</li>
</ul>
<p>Popular architectures: <a href="https://en.wikipedia.org/wiki/Perceptron">Perceptron</a>, <a href="https://en.wikipedia.org/wiki/Convolutional_neural_network">Convolutional Network</a> (CNN), <a href="https://en.wikipedia.org/wiki/Recurrent_neural_network">Recurrent Networks</a> (RNN), <a href="https://en.wikipedia.org/wiki/Autoencoder">Autoencoders</a></p>

</div>

<div>
<p>If no one has ever tried to explain neural networks to you using "human brain" analogies, you're happy. Tell me your secret. But first, let me explain it the way I like.</p>

<p>Any neural network is basically a collection of <strong>neurons</strong> and <strong>connections</strong> between them. <strong>Neuron</strong> is a function with a bunch of inputs and one output. Its task is to take all numbers from its input, perform a function on them and send the result to the output.</p>

<p>Here is an example of a simple but useful in real life neuron: sum up all numbers from the inputs and if that sum is bigger than N — give 1 as a result. Otherwise — zero.</p>

<p><strong>Connections</strong> are like channels between neurons. They connect outputs of one neuron with the inputs of another so they can send digits to each other. Each connection has only one parameter — weight. It's like a connection strength for a signal. When the number 10 passes through a connection with a weight 0.5 it turns into 5.</p>

<p>These weights tell the neuron to respond more to one input and less to another. Weights are adjusted when training — that's how the network learns. Basically, that's all there is to it.</p>

<p>To prevent the network from falling into anarchy, the neurons are linked by layers, not randomly. Within a layer neurons are not connected, but they are connected to neurons of the next and previous layers. Data in the network goes strictly in one direction — from the inputs of the first layer to the outputs of the last.</p>

<p>If you throw in a sufficient number of layers and put the weights correctly, you will get the following: by applying to the input, say, the image of handwritten digit 4, black pixels activate the associated neurons, they activate the next layers, and so on and on, until it finally lights up the exit in charge of the four. The result is achieved.</p>

<p>When doing real-life programming nobody is writing neurons and connections. Instead, everything is represented as matrices and calculated based on matrix multiplication for better performance. My favourite video on this and its sequel below describe the whole process in an easily digestible way using the example of recognizing hand-written digits. Watch them if you want to figure this out.</p>
<div>
<p>A network that has multiple layers that have connections between every neuron is called a <a href="https://en.wikipedia.org/wiki/Multilayer_perceptron">perceptron</a> (MLP) and considered the simplest architecture for a novice. I didn't see it used for solving tasks in production.</p>
</div>

<p>After we constructed a network, our task is to assign proper ways so neurons will react correctly to incoming signals. Now is the time to remember that we have data that is samples of 'inputs' and proper 'outputs'. We will be showing our network a drawing of the same digit 4 and tell it 'adapt your weights so whenever you see this input your output would emit 4'.</p>

<p>To start with, all weights are assigned randomly. After we show it a digit it emits a random answer because the weights are not correct yet, and we compare how much this result differs from the right one. Then we start traversing network backward from outputs to inputs and tell every neuron 'hey, you did activate here but you did a terrible job and everything went south from here downwards, let's keep less attention to this connection and more of that one, mkay?'.</p>

<p>After hundreds of thousands of such cycles of 'infer-check-punish', there is a hope that the weights are corrected and act as intended. The science name for this approach is <a href="https://en.wikipedia.org/wiki/Backpropagation">Backpropagation</a>, or a 'method of backpropagating an error'. Funny thing it took twenty years to come up with this method. Before this we still taught neural networks somehow.</p>

<p>My second favorite vid is describing this process in depth, but it's still very accessible.</p>

<p>A well trained neural network can fake the work of any of the algorithms described in this chapter (and frequently works more precisely). This universality is what made them widely popular. <em>Finally we have an architecture of human brain</em> they said <em>we just need to assemble lots of layers and teach them on any possible data</em> they hoped. Then the first <a href="https://vas3k.com/blog/machine_learning/(https://en.wikipedia.org/wiki/AI_winter">AI winter</a>) started, then it thawed, and then another wave of disappointment hit.</p>

<p>It turned out networks with a large number of layers required computation power unimaginable at that time. Nowadays any gamer PC with geforces outperforms the datacenters of that time. So people didn't have any hope then to acquire computation power like that and neural networks were a huge bummer.</p>

<p>And then ten years ago deep learning rose.</p>

<p>In 2012 convolutional neural networks acquired an <a href="https://habr.com/post/183380/">overwhelming victory in ImageNet competition</a> that made the world <em>suddenly remember</em> about methods of deep learning described in the ancient 90s. Now we have video cards!</p>

<p>Differences of deep learning from classical neural networks were in new methods of training that could handle bigger networks. Nowadays only theoretics would try to divide which learning to consider deep and not so deep. And we, as practitioners are using popular 'deep' libraries like <a href="https://keras.io">Keras</a>, <a href="https://github.com/tensorflow/tensorflow">TensorFlow</a> &amp; <a href="https://pytorch.org">PyTorch</a> even when we build a mini-network with five layers. Just because it's better suited than all the tools that came before. And we just call them neural networks.</p>

<p>I'll tell about two main kinds nowadays.</p>
<h3>Convolutional Neural Networks (CNN)</h3>

<p>Convolutional neural networks are all the rage right now. They are used to search for objects on photos and in videos, face recognition, style transfer, generating and enhancing images, creating effects like slow-mo and improving image quality. Nowadays CNNs are used in all the cases that involve pictures and videos. Even in your iPhone several of these networks are going through your nudes to detect objects in those. If there <em>is something to detect</em>, heh.</p>

<div>
<p>Image above is a result produced by <a href="https://github.com/facebookresearch/Detectron">Detectron</a> that was recently open-sourced by Facebook</p>
</div>

<p>A problem with images was always the difficulty of extracting features out of them. You can split text by sentences, lookup words' attributes in specialized vocabularies, etc. But images had to be labeled manually to teach the machine where cat ears or tails were in this specific image. This approach got the name 'handcrafting features' and used to be used almost by everyone.</p>

<p>There are lots of issues with the handcrafting.</p>

<p>First of all, if a cat had its ears down or turned away from the camera: you are in trouble, the neural network won't see a thing.</p>

<p>Secondly, try naming on the spot 10 different features that distinguish cats from other animals. I for one couldn't do it, but when I see a black blob rushing past me at night — even if I only see it in the corner of my eye — I would definitely tell a cat from a rat. Because people don't look only at ear form or leg count and account lots of different features they don't even think about. And thus cannot explain it to the machine.</p>

<p>So it means the machine needs to learn such features on its own, building on top of basic lines. We'll do the following: first, we divide the whole image into 8x8 pixel blocks and assign to each a type of dominant line – either horizontal [-], vertical [|] or one of the diagonals [/]. It can also be that several would be highly visible — this happens and we are not always absolutely confident.</p>

<p>Output would be several tables of sticks that are in fact the simplest features representing objects edges on the image. They are images on their own but built out of sticks. So we can once again take a block of 8x8 and see how they match together. And again and again…</p>

<p>This operation is called convolution, which gave the name for the method. Convolution can be represented as a layer of a neural network, because each neuron can act as any function.</p>

<p>When we feed our neural network with lots of photos of cats it automatically assigns bigger weights to those combinations of sticks it saw the most frequently. It doesn't care whether it was a straight line of a cat's back or a geometrically complicated object like a cat's face, something will be highly activating.</p>

<p>As the output, we would put a simple perceptron which will look at the most activated combinations and based on that differentiate cats from dogs.</p>

<p>The beauty of this idea is that we have a neural net that searches for the most distinctive features of the objects on its own. We don't need to pick them manually. We can feed it any amount of images of any object just by googling billions of images with it and our net will create feature maps from sticks and learn to differentiate any object on its own.</p>

<p>For this I even have a handy unfunny joke:</p>
<blockquote>
<p>Give your neural net a fish and it will be able to detect fish for the rest of its life. Give your neural net a fishing rod and it will be able to detect fishing rods for the rest of its life…</p>
</blockquote>
<h3>Recurrent Neural Networks (RNN)</h3>

<p>The second most popular architecture today. Recurrent networks gave us useful things like neural machine translation (<a href="https://vas3k.com/blog/machine_translation/">here is my post about it</a>), speech recognition and voice synthesis in smart assistants. RNNs are the best for sequential data like voice, text or music.</p>

<p>Remember Microsoft Sam, the old-school speech synthesizer from Windows XP? That funny guy builds words letter by letter, trying to glue them up together. Now, look at Amazon Alexa or Assistant from Google. They don't only say the words clearly, they even place the right accents!</p>
<p>
    VIDEO
    <small>Neural Net is trying to speak</small>
</p>
<p>All because modern voice assistants are trained to speak not letter by letter, but on whole phrases at once. We can take a bunch of voiced texts and train a neural network to generate an audio-sequence closest to the original speech.</p>

<p>In other words, we use text as input and its audio as the desired output. We ask a neural network to generate some audio for the given text, then compare it with the original, correct errors and try to get as close as possible to ideal.</p>

<p>Sounds like a classical learning process. Even a perceptron is suitable for this. But how should we define its outputs? Firing one particular output for each possible phrase is not an option — obviously.</p>

<p>Here we'll be helped by the fact that text, speech or music are sequences. They consist of consecutive units like syllables. They all sound unique but depend on previous ones. Lose this connection and you get dubstep.</p>

<p>We can train the perceptron to generate these unique sounds, but how will it remember previous answers? So the idea is to add memory to each neuron and use it as an additional input on the next run. A neuron could make a note for itself - hey, we had a vowel here, the next sound should sound higher (it's a very simplified example).</p>

<p>That's how recurrent networks appeared.</p>

<p>This approach had one huge problem - when all neurons remembered their past results, the number of connections in the network became so huge that it was technically impossible to adjust all the weights.</p>

<p>When a neural network can't forget, it can't learn new things (people have the same flaw).</p>

<p>The first decision was simple: limit the neuron memory. Let's say, to memorize no more than 5 recent results. But it broke the whole idea.</p>

<p>A much better approach came later: to use special cells, similar to computer memory. Each cell can record a number, read it or reset it. They were called long and short-term memory (LSTM) cells.</p>

<p>Now, when a neuron needs to set a reminder, it puts a flag in that cell. Like "it was a consonant in a word, next time use different pronunciation rules". When the flag is no longer needed, the cells are reset, leaving only the “long-term” connections of the classical perceptron. In other words, the network is trained not only to learn weights but also to set these reminders.</p>

<p>Simple, but it works!</p>
<p>
    VIDEO
    <small>CNN + RNN = Fake Obama</small>
</p>
<p>You can take speech samples from anywhere. BuzzFeed, for example, took Obama's speeches and trained a neural network to imitate his voice. As you see, audio synthesis is already a simple task. Video still has issues, but it's a question of time.</p>

<hr>
<p>There are many more network architectures in the wild. I recommend a good article called <a href="http://www.asimovinstitute.org/neural-network-zoo/">Neural Network Zoo</a>, where almost all types of neural networks are collected and briefly explained.</p>

<p></p>

</div>

<h1>The End: when the war with the machines?</h1>
<div>
<p>The main problem here is that the question <em>"when will the machines become smarter than us and enslave everyone?"</em> is initially wrong. There are too many hidden conditions in it.</p>

<p>We say "become smarter than us" like we mean that <strong>there is a certain unified scale of intelligence</strong>. The top of which is a human, dogs are a bit lower, and stupid pigeons are hanging around at the very bottom.</p>

<p>That's wrong.</p>

<p>If this were the case, every human must beat animals in everything but it's not true. The average squirrel can remember a thousand hidden places with nuts — I can't even remember where are my keys.</p>

<p>So intelligence is a set of different skills, not a single measurable value? Or is remembering nuts stashed locations not included in intelligence?</p>

<p>An even more interesting question for me - <strong>why do we believe that the human brain possibilities are limited?</strong> There are many popular graphs on the Internet, where the technological progress is drawn as an exponent and the human possibilities are constant. But is it?</p>

<p>Ok, multiply 1680 by 950 right now in your mind. I know you won't even try, lazy bastards. But give you a calculator — you'll do it in two seconds. Does this mean that the calculator just expanded the capabilities of your brain?</p>

<p>If yes, can I continue to expand them with other machines? Like, use notes in my phone to not to remember a shitload of data? Oh, seems like I'm doing it right now. I'm expanding the capabilities of my brain with the machines.</p>

<p>Think about it. Thanks for reading.</p>

</div>

        </div>

        

</div><br>
<br>
<br>
<br>
<br>
<br>
via vas3k.com https://vas3k.com<br>
<br>
February 10, 2019 at 11:37AM
